docs/articles/rag_hr_chatbot.md

```diff
@@ -182,108 +193,133 @@ source.put([text_df])
+Et voila! Now all results are fresh; doc 0 and other outdated results are completely removed from the top ten results. 
 
-Note also that, importantly, the same weights (including our new recency_weight=0.25) produce relevant results for queries about maternity leave—a topic discussed only in the older HR document (2023). This older material does not get overruled by more recent but not relevant documents.
+Still, some of our top results have low **usefulness scores**. Documents 24 and 35, for example, both in our top ten results, despite their radically different usefulness scores (0.83 vs. 0.04). Document 24 (2024) softens the review of info security compliance requirement from document 35 (2023)'s "in all cases" to "wherever applicable". To eliminate the document that employees find useless (unsurprisingly, given the text), we have to upweight more than just recency or relevancy, which fails to sufficiently differentiate between the two documents. This is where the `usefulness_weight` comes in.
 
 ```python
-normal_recency_result = app.query(
-   knowledgebase_query,
-   relevance_weight=1,
-   recency_weight=0.25,
-   search_query="What are the companies terms for maternal leave?",
-   limit=TOP_N,
+normal_recency_usefulness_result = app.query(
+    knowledgebase_query,
+    relevance_weight=1,
+    recency_weight=0.4,
+    usefulness_weight=0.25,
+    search_query=initial_query_text,
+    limit=TOP_N,
 )
 
+normal_recency_usefulness_result_df = present_result(normal_recency_usefulness_result)
+normal_recency_usefulness_result_df
+```
+
+![Results - usefulness weighted](../assets/use_cases/rag_hr_chatbot/results-usefulness.png)
+
+Weighting usefulness eradicates all irrelevant docs from our top ten results!
+
+It is also important to see that the same weights produce relevant results for queries about maternity leave — a topic only discussed in the older HR document. It should not get overruled by recent but not relevant documents....
+
+Note also, importantly, that if we query "maternity leave" (below), the same weights (as above) produce relevant results for queries about maternity leave — a topic discussed only in the older HR document (2023). With the proper weighting, this older material appropriately does *not* get overruled by more recent but *irrelevant* documents.
 
-maternity_result = present_result(normal_recency_result)
-maternity_result
+```python
+maternity_result = app.query(
+    knowledgebase_query,
+    relevance_weight=1,
+    recency_weight=0.4,
+    usefulness_weight=0.25,
+    search_query="What are the companies terms for maternal leave?",
+    limit=TOP_N,
+)
+
+maternity_result_df = present_result(maternity_result)
+maternity_result_df
 ```
 
-![Results 4](../assets/use_cases/rag_hr_chatbot/result4.png)
+![Results 4](../assets/use_cases/rag_hr_chatbot/results4.png)
```

Comment:
how to interpret extremely low usefulness scores of our results?

Reply:
I don't think that is the case - around 0.6 is ok usefulness score